File: 백준/Gold/14502. 연구소/연구소.cc
#include <bits/stdc++.h>
using namespace std;
using pii = pair<int, int>;
const int MAX = 8;
int N, M;
int MAP[MAX][MAX], TMP[MAX][MAX];
vector<pii> virus, empty_spaces;

int dx[] = { 0, 0, 1, -1 };
int dy[] = { 1, -1, 0, 0 };

void spreadVirus() {
    queue<pii> q;
    for (auto v : virus) q.push(v);

    while (!q.empty()) {
        auto [x, y] = q.front();
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i], ny = y + dy[i];
            if (nx >= 0 && ny >= 0 && nx < N && ny < M && TMP[nx][ny] == 0) {
                TMP[nx][ny] = 2;
                q.push({ nx, ny });
            }
        }
    }
}

int getSafeZone() {
    int safe = 0;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++)
            if (TMP[i][j] == 0) safe++;
    return safe;
}

int solve() {
    int max_safe_zone = 0;
    vector<int> perm(empty_spaces.size(), 0);
	fill(perm.end() - 3, perm.end(), 1); // [0,0,0,...1,1,1]

    do {
        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++)
                TMP[i][j] = MAP[i][j];

        for (int i = 0; i < empty_spaces.size(); i++)
            if (perm[i]) TMP[empty_spaces[i].first][empty_spaces[i].second] = 1;

        spreadVirus();
        max_safe_zone = max(max_safe_zone, getSafeZone());

    } while (next_permutation(perm.begin(), perm.end()));

    return max_safe_zone;
}

int main() {
    cin.tie(0)->sync_with_stdio(0);
    cin >> N >> M;
    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++) {
            cin >> MAP[i][j];
            if (MAP[i][j] == 0) empty_spaces.push_back({ i, j });
            if (MAP[i][j] == 2) virus.push_back({ i, j });
        }

    cout << solve();
}

File: 백준/Gold/14502. 연구소/연구소.py
import sys
from itertools import permutations
from collections import deque

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

N, M = map(int, sys.stdin.readline().split())
MAP = []
empty_spaces = []
virus = []

for i in range(N):
    MAP.append(list(map(int, sys.stdin.readline().split()))) 
    for j in range(M):
        if MAP[i][j] == 0:
            empty_spaces.append((i, j))
        elif MAP[i][j] == 2:
            virus.append((i, j))
            
def spread_virus(TMP):
    q = deque(virus)
    
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < N and 0 <= ny < M and TMP[nx][ny] == 0:
                TMP[nx][ny] = 2
                q.append((nx, ny))

def get_safe_zone(TMP):
    return sum(row.count(0) for row in TMP)

def solve():
    """ 모든 벽 배치 경우를 확인하여 최대 안전 영역 찾기 """
    max_safe_zone = 0

    for walls in permutations(empty_spaces, 3):
        TMP = [row[:] for row in MAP]

        for x, y in walls:
            TMP[x][y] = 1

        spread_virus(TMP)
        max_safe_zone = max(max_safe_zone, get_safe_zone(TMP))

    return max_safe_zone

print(solve())